Return Message Information to Callbacks

obsoleted opened the pull request:
This is a **proposal** for one way we might return message information to
the application registered callbacks/listeners. I definitely want folks feedback before this goes in.
- This method returns a message information object as a 2nd parameter to the application callbacks. 
- The msg information is limited to a few fields now but this can easily changed
- The advantage of this implementation is that it doesn't require changes for our existing apps. 
- This first pass only touches success callbacks that were receiving arguments. When we decide on the way we want to do this we should then make it generic and return it for all message backed callbacks as well as for errors.

The change is spread across two layers:

Native/Proxy layer:
- Where this layer used to just return an array of arguments for the
  callback it now returns an array of two values. The first value is a
  dictionary structure representing the message that resulted in the
  callback. The second value is the argument array that we used to
  return

AllJoyn.js layer:
- Because of the changes in the native/proxy layer we now wrap all
  callback methods that receive this new message info & argument array.
  The wrapper splits up the message info and calls the callback with the
  argument array in the first parameter (parity with existing behavior)
  and the msg info object as the 2nd parameter.

Here is a rough example of how this change might be used:
In the chat sample we can have code like this:

``` javascript
        var chatMessageHandler = function(response, msgInfo) {
          if (channelsModel.currentChannel == null) return;
          var message = new Message(response[0], msgInfo.sender);
          channelsModel.currentChannel.messages.push(message);
          broadcastOnRootScope('newMessage', message);
        }
```

Which will have a result like this:
![chat_with_sender](https://cloud.githubusercontent.com/assets/666628/6312411/30e3b72c-b92c-11e4-9631-032c6f37b893.jpg)

obsoleted committed:
Return Message Information to Callbacks

The change is spread across two layers:

Native/Proxy layer:
- Where this layer used to just return an array of arguments for the
  callback it now returns an array of two values. The first value is a
  dictionary structure representing the message that resulted in the
  callback. The second value is the argument array that we used to
  return

AllJoyn.js layer:
- Because of the changes in the native/proxy layer we now wrap all
  callback methods that receive this new message info & argument array.
  The wrapper creates a new message object that contains both the
  unmarshalled arguments and the message information.

An example of what will be returned to an application callback:
{
    sender: ':slkdf.3', // From the message information
    signature: '(nnn)', // From the message information
    arguments: [[1,2,3]], // From arguments
}

Additionally this change refactors the addInterfacesListener API to
avoid using an exec call and instead use setSignalRule and addListener.

Update addAdvertisedNameListener api to use message object

Signals do not receive message callback

obsoleted commented:
@vjrantal @phongcao Feel free to take a look and let me know your thoughts.

For example, another way we might want to expose the information is to refactor the API to invoke the callback with one parameter that had different properties for the message information and arguments. 
So something like this:

``` javascript
{
  sender: ':asdfsdfa.1',
  iface: 'org.example.foo',
  arguments: [ 'arg1', [1, 2, ,3], 's', 'variantstringvalue'],
}
```

Of course, such a option would break the samples. But we don't have that many now :smile: 

We could also return class (e.g. new AllJoyn.Message(...)) and give users the ability to extend it. Hard to say how useful that might be though.

phongcao commented:
For our common APIs, I prefer to have one object containing all message info (properties + arguments) like the above example. However, it's your call who are experts on this. :)

vjrantal commented:
I would also tend to lean towards only one object, which contains the arguments. I would keep it a dictionary with keys like sender, iface, arguments etc. just like in the example above.

@obsoleted: would you have the courage and willingness to change it like that and update the samples we have? :)

Once done, we could also include an automated test to verify the object we return.

obsoleted commented:
Ok. I will update to return one object that contains arguments and other message properties.

I will also send out PRs for the other org samples at the same time.

And add tests if possible :smile:

obsoleted commented:
Updated PR includes the following:
- return more message information within a single object that also includes the unmarshalled arguments to the callback
- includes updates to the addAdvertisedNameListener and addInterfacesListener functions to include the original message in the objects these callbacks return
- The implementation for addInterfacesListener was also updated to rely on other higher level APIs and avoid the need for an exec call of the same name

Example callback parameter objects below:
Method callback:

``` JavaScript
{
    sender: ':lskdj.1',
    signature: 'sanv',
    arguments: [ ['stringval', [1,2,3], 'n', 3],
}
```

addInterfacesListener callback:

``` JavaScript
{
   version: 1,
   port: 25,
   message: messageObject, // See above
   objects: [ { path: '/foo', interfaces: ['com.example.iface1', 'com.example.iface2']  } ]
}
```

addAdvertisedNameListener callback{

``` JavaScript
{
    name: 'com.example.chat.sample.Foo',
    message: messageObject, //see above
}
```

obsoleted committed:
Move about announcement properties under a properties field

- Instead of risking conflicts between existing fields in the
  announcement (e.g., port)
